### ISSUE TITLE
The dreaded != w and related error reporting

### ISSUE THREAD

**Issue by GoogleCodeExporter** [agda/agda#603]
```
In the following program (written by Rotsor on the IRC channel)

module w where

data ℕ : Set where
  zero : ℕ
  suc  : (n : ℕ) → ℕ

data IsSuc : ℕ → Set where
  iss : (n : ℕ) → IsSuc (suc n)

data ⊤ {A : Set} : A → Set where
  tt : ∀ x → ⊤ x

qwe : (n : ℕ) → ⊤ (iss n)
qwe n with suc n
... | sn = {!!}

the with block itself generates a type error. There is no need to pattern match 
on the result, or anything like that. The type error refers to a mysterious != 
w, where the w is a generated name in the code the with block desugars to. 
Recently, the error message was amended to show the generated code (before it 
used to just say w without telling you where it came from), but that still 
isn't particularly helpful.

Ulf said on IRC that this kind of issue usually arises when you need to 
abstract over more things in the with block in question, so why not amend the 
error message to say something to that effect? In the above example, adding 
"iss n" to the with block fixes it. I guess then the question becomes whether 
Agda can figure out what's missing enough to make a suggestion about it. In 
more complicated examples, there might be more than one way to solve the issue, 
but it'd be nice to provide a suggestion and a user-friendly error message 
along the lines of:

Unable to construct a with block, probably because more expressions need to be 
added to the block. Perhaps adding "iss n" would help?

The last bit could be omitted if it's hard to determine what would help.

In a similar vein, when matching on a constructor that refines indices, and 
those indices are not in scope, the error message currently says something like 
"could not figure out how to unify indices [a, b, c] with [...]". Maybe that 
message could be adjusted in a similar manner? And it seems like in that case, 
making a constructive suggestion for fixing it would be a lot easier than the 
!= w case.

Anyway, I don't have a clear solution here, but it's very frustrating when my 
goal normalization is blocked on an expression, but adding it to a with block 
gives me a type error. I'd be interested to see if anyone else has better 
suggestions on what to do here.
```

Original issue reported on code.google.com by `pumpkingod@gmail.com` on 4 Apr 2012 at 6:17

**Comment by jespercockx** [comment]
Putting this in the `icebox` until someone comes up with a concrete suggestion how to improve the error message.
